doc/man/man1/ocpiadmin.1
.\"     Title: ocpiadmin
.\"    Author: [FIXME: author] [see http://www.docbook.org/tdg5/en/html/author]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 12/02/2020
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "OCPIADMIN" "1" "12/02/2020" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ocpiadmin \- command\-line tool for managing an OpenCPI installation
.SH "SYNOPSIS"
.sp
\fBocpiadmin\fR [\fI<options>\fR] \fI<verb>\fR \fI<noun>\fR [\fI<name>\fR [\fI<argument>\fR\&... ]]
.SH "DESCRIPTION"
.sp
The \fBocpiadmin(1)\fR command\-line tool is used to perform installation and deployment tasks that are carried out on an OpenCPI installation after the basic installation procedure is completed to prepare OpenCPI for use on embedded systems\&. These tasks include:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Building platforms and simulators that are intended as targets for OpenCPI development that have not been automatically built as part of the basic OpenCPI installation\&. Building a platform enables OpenCPI assets for that platform to be developed in OpenCPI projects\&. It also enables the platform to be integrated into system boot kits \- for example, on bootable SD cards \- to be created for embedded systems that use them\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Creating a bootable media directory tree (\fISD card contents\fR) for OpenCPI required by the software (RCC) and hardware (HDL/FPGA) platforms that comprise the embedded system\&. The bootable media directory tree for a particular software\-hardware platform configuration can then be written to the manufacturer\(cqs SD card to enable booting the embedded system for OpenCPI\&.
.RE
.sp
For more information about these tasks and how they relate to enabling OpenCPI development on embedded systems, see the \fIOpenCPI Installation Guide\fR\&.
.sp
The general usage concept for \fBocpiadmin\fR is: perform the \fI<verb>\fR operation on the \fI<noun>\fR whose name is \fI<name>\fR\&.
.SH "VERBS"
.sp
The verbs supported by \fBocpiadmin\fR are:
.PP
\fBdeploy\fR
.RS 4
Create a bootable media directory tree for the named RCC (software) platform on the named HDL (FPGA) platform\&. The specified platforms must have been previously built\&.
.RE
.PP
\fBinstall\fR
.RS 4
Prepare the OpenCPI installation for using the named RCC or HDL platform or simulator\&.
.RE
.SH "NOUNS"
.sp
The nouns supported by \fBocpiadmin\fR are:
.PP
\fBplatform\fR
.RS 4
A particular type of processing hardware and/or software that can host a container for executing OpenCPI workers\&.
.RE
.SH "OPTIONS"
.sp
The following options can be specified for the \fBinstall\fR verb only:
.PP
\fB\-\-git\-revision=\fR\fI<git\-rev>\fR, \fB\-g\fR \fI<git\-rev>\fR
.RS 4
Specifies the branch, tag or other valid
\fBgit\fR
revision to check out after downloading/cloning the project specified in the
\fB\-\-package\-id\fR
(\fB\-p\fR) option (described below)\&. This option can be used with the
\fB\-\-package\-id\fR
option to install a different
\fBgit\fR
revision of a project from the one currently checked out\&. If this option is not used, the
\fBgit\fR
revision for the specified package\-ID defaults to the
\fBgit\fR
revision of the OpenCPI framework repository\&.
.RE
.PP
\fB\-\-package\-id=\fR\fI<package\-id>\fR, \fB\-p\fR \fI<package\-id>\fR
.RS 4
Specifies the OpenCPI package\-ID of the project that provides the support for the platform\&. This option is only required when support for the platform is provided in an
\fIOSP\fR
(\fIOpenCPI System support Platform\fR) instead of in a built\-in OpenCPI project\&. The default format for a package\-ID is
\fI<package\-prefix>\fR\fB\&.\fR\fI<package\-name>\fR; for example,
\fBocpi\&.osp\&.e3xx\fR,
\fBocpi\&.osp\&.plutosdr\fR\&. See the chapter on package\-IDs in the
\fIOpenCPI Component Development Guide\fR
for details\&.
.RE
.PP
\fB\-\-url=\fR\fI<url>\fR, \fB\-u\fR \fI<url>\fR
.RS 4
Specifies the internet URL where the project resides\&. This option can be used with the
\fB\-\-package\-id\fR
(\fB\-p\fR) option (described above) to specify the location of a project that is not hosted on the OpenCPI Gitlab site\&. If this option is not used, the URL defaults to
\fBhttps://gitlab\&.com/opencpi/osps/\fR\fI<package\-id>\fR\fB\&.git\fR, where
\fI<package\-id>\fR
is the one specified in the
\fB\-\-package\-id=\fR
option\&. For example, if
\fI<package\-id>\fR
is
\fBocpi\&.osp\&.e3xx\fR, the URL is
\fBhttps://gitlab\&.com/opencpi/osps/ocpi\&.osps\&.e3xx\&.git\fR\&.
.RE
.SH "OPERATIONS"
.sp
The command syntax for the \fBinstall platform\fR operation is:
.sp
\fBocpiadmin\fR \fI<options>\fR \fBinstall platform\fR \fI<platform>\fR
.sp
where \fI<platform>\fR is the name of an RCC platform, HDL platform or HDL simulator to be built for OpenCPI\&. If necessary, the operation downloads the platform\(cqs OSP into the \fBprojects/osps/\fR directory in the OpenCPI installation directory and then builds the built\-in projects and the platform\(cqs OSP (if any) for the specified platform\&. If the specified platform is an HDL platform, the operation also builds a single test artifact/bitstream: the \fBtestbias\fR assembly in the \fBassets\fR project\&. After the operation is performed, any asset in any project can be built for the specified platform\&.
.sp
The command syntax for the \fBdeploy platform\fR operation is:
.sp
\fBocpiadmin deploy platform\fR \fI<rcc\-platform> <hdl\-platform>\fR
.sp
where \fI<rcc\-platform>\fR and \fI<hdl\-platform>\fR are the names of the software (RCC) and hardware/FPGA (HDL) platforms for which the bootable media should be created\&. The operation creates the directory \fBcdk/\fR\fI<hdl\-platform>\fR\fB/sdcard\-\fR\fI<rcc\-platform>\fR, whose contents can be copied to an SD card\&.
.sp
Keywords for specifying platform names are given in tables in the \fIOpenCPI Installation Guide\fR\&.
.SH "EXAMPLES"
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  1." 4.2
.\}
Build the Xilinx Xsim (\fBxsim\fR) HDL simulator platform for OpenCPI:
.sp
.if n \{\
.RS 4
.\}
.nf
ocpiadmin install platform xsim
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  2." 4.2
.\}
Download and build the Ettus USRP E310 (\fBe31x\fR) HDL platform for OpenCPI:
.sp
.if n \{\
.RS 4
.\}
.nf
 ocpiadmin \-\-package\-id=ocpi\&.osp\&.e3xx install platform e31x
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 3.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  3." 4.2
.\}
Download and build the Analog Devices Pluto Linux 0\&.32 (\fBadi_plutosdr0_32\fR) embedded software (RCC) platform for OpenCPI:
.sp
.if n \{\
.RS 4
.\}
.nf
ocpiadmin \-\-package\-id=ocpi\&.osp\&.plutosdr install platform adi_plutosdr0_32
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 4.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  4." 4.2
.\}
Build the Analog Devices ADALM\-PLUTO (\fBplutosdr\fR) platform (you have already downloaded the OSP that supports this platform if you performed the previous example):
.sp
.if n \{\
.RS 4
.\}
.nf
ocpiadmin install platform plutosdr
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 5.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  5." 4.2
.\}
Download the third\-party OSP that provides the
\fBmyplat\fR
platform from the Github site and build the platform:
.sp
.if n \{\
.RS 4
.\}
.nf
ocpiadmin \-\-package\-id=com\&.myco\&.myplat \-\-url=https://github\&.com/myco/myplat\&.git install platform myplat
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 6.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  6." 4.2
.\}
Download and build the latest "development"
\fBgit\fR
revision of the
\fBplutosdr\fR
platform from a working branch using an earlier version:
.sp
.if n \{\
.RS 4
.\}
.nf
ocpiadmin \-\-package\-id=ocpi\&.osp\&.plutosdr \-\-git\-revision=develop install platform plutosdr
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 7.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  7." 4.2
.\}
Create a bootable media directory tree for booting the RCC platform
\fBxilinx19_2_aarch32\fR
on the HDL platform
\fBzed\fR\&. This operation creates the directory
\fBcdk/zed/sdcard\-xilinx19_2_aarch32/\fR, whose contents can be copied to an SD card:
.sp
.if n \{\
.RS 4
.\}
.nf
ocpiadmin deploy platform xilinx19_2_aarch32 zed
.fi
.if n \{\
.RE
.\}
.RE
.SH "BUGS"
.sp
See https://www\&.opencpi\&.org/report\-defects
.SH "RESOURCES"
.sp
See the main web site: https://www\&.opencpi\&.org
.SH "SEE ALSO"
.sp
ocpidev(1) ocpidev\-platform(1) ocpidev\-project(1) ocpidev\-worker(1)
.SH "COPYING"
.sp
Copyright (C) 2020 OpenCPI www\&.opencpi\&.org\&. OpenCPI is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version\&.


CHANGELOG.md
# [v2.1.0-beta.1](https://gitlab.com/opencpi/opencpi/-/compare/v2.0.1...v2.1.0-beta.1) (2020-12-18)

Changes/additions since [OpenCPI Release v2.0.1](https://gitlab.com/opencpi/opencpi/-/releases/v2.0.1)

### New Features
- **runtime**: slaves are not automatically included in the application without mentioning them. (!428)(22c9956b)
- **tools**: add convenience script (scripts/install-gnuradio.sh) to install OpenCPI's modified version of GNU Radio 3.7. (!283)(8ad0cd5f)

### Enhancements
- **app**: enable bbloopback mode for FSK application on matchstiq z1. (!422)(2c2ebb0d)
- **devops**: add OSP support to yaml-generator.py. (!382)(bf7a9c77)
- **devops**: implement child pipelines for cross-platform builds/tests. (!382)(bf7a9c77)
- **devops**: implement grandchild pipelines so that more jobs can be dynamically generated. (!431)(c249580e)
- **devops**: fix downstream pipelines failing due to environment variables not being set correctly. (!437)(72adeca8)
- **osp**: add support for zcu102. (!407)(444d760c)
- **protocols**: update ComplexShortWithMetadata-prot.xml to use a sequence length of 4096 instead of 4092 for the iq argument of the samples operation. (!427)(34d0d193)
- **runtime**: port delegation from proxies to slaves is now implemented. (!421)(b718ff61)
- **runtime**: generic DRC helper class established in an API header OcpiDrcProxyApi.hh. (!421)(b718ff61)
- **runtime**: update zed/fmcomms2 platform with new drc. (!433)(d494dad9)
- **tools**: install-gnuradio.sh: add GNU Radio 3.8 support. (!401)(4c92ed80)
- **tools**: add missing fields to the opencpi GRC blocks. (!425)(a40d1a90)
- **tools**: ocpigen: add XML parsing for project assets. (!434)(30acd3d2)

### Bug Fixes
- **runtime**: fix ocpiremote deploy not deploying boot files to correct remote directory. (!416)(56b316af)
- **tools**: include bitstream argument in "ocpiremote restart" to prevent errors being thrown. (!371)(83566d76)
- **tools**: improve gnuradio 3.8 install. (!405)(a18a710c)

### Miscellaneous
- **doc,runtime,tests,tools**: remove remaining CentOS 6 references and supporting code. (!403)(3b9c943b)
- **comp**: make "socket_write.rcc" portable. (!418)(4755f36a)
- **hdl base**: add ADI "mykonos" library as a prerequisite. (!385)(f2ec19c5)
- **tools**: remove remaining BSV references and code from ocpigen. (!414)(8f73efd2)

# [v2.0.1](https://gitlab.com/opencpi/opencpi/-/compare/v2.0.0...v2.0.1) (2020-11-18)

Changes/additions since [OpenCPI Release v2.0.0](https://gitlab.com/opencpi/opencpi/-/releases/v2.0.0)

### Enhancements
- **osp**: update/refactor the DRC support libraries to be sharable among proxies, and make pluto an example of using them. (!393)(df56b22b)
- **tools**: doc site: include OSP documentation starting with version 1.7.1. (!402)(73a409c4)
- **tools**: doc site: add man page link to a release's index page if man pages exist. (!402)(73a409c4)

### Bug Fixes
- **doc,tools**: prevent libreoffice tracked changes from showing up in rendered pdfs. (!392)(a7cb3110)
- **hdl base**: add IDLE state to state machine. (!404)(477404a2)
- **hdl base**: fix bug with statemachine. (!406)(d7fca60c)
- **runtime**: add new way of calling no-os ad9361 library. (!420)(01e79568)
- **tools**: fix bugs for clock adapter insertion. (!383)(2632c226)
- **tools**: fix ubuntu check scripts to not error on non-ubuntu platforms. (!398)(044e4db1)
- **tools**: fix bugs for clock adapter insertion. (!399)(99e89a32)
- **tools**: doc site: fix incorrect "latest" release labeling. (!402)(73a409c4)
- **tools**: fix "ocpigen" processing of 64-bit initial values for 64-bit HDL worker parameters. (!417)(977c228e)

### Miscellaneous
- **comp**: move "zero_padding" component to "inactive" project. (!419)(64f4f058)
- **doc,runtime,tests,tools**: remove remaining CentOS 6 references and supporting code. (!403)(3b9c943b)

# [v2.0.0](https://gitlab.com/opencpi/opencpi/-/compare/v2.0.0-rc.2...v2.0.0) (2020-10-06)

Changes/additions since [OpenCPI Release v2.0.0-rc.2](https://gitlab.com/opencpi/opencpi/-/releases/v2.0.0-rc.2)

### New Features
- **tools**: add dockerfile for docker image used during GRCon20 workshops. (!386)(94e17772)

### Enhancements
- **app**: include the demo app from tutorial 1 in the tutorial project for completeness. (!373)(1566b9e9)
- **comp**: Add timed sample protocols. (!370)(90c31833)
- **comp**: add rcc and hdl workers for the tutorial 1 workers so we have both rcc and hdl of all of them. (!373)(1566b9e9)
- **comp**: cleanup another tutorial worker, `peak_detector.rcc`, for tutorial 2. (!373)(1566b9e9)
- **doc**: add files for the `timestamper_scdcd` datasheet to be compiled. (!357)(37931d3b)
- **runtime**: to be more consistent, enable "worker" attribute in app instances to include the authoring model suffix. (!373)(1566b9e9)
- **runtime**: allow "worker" variable in selection expressions. (!373)(1566b9e9)
- **runtime**: when there are no slaves indicated for a proxy instance (like from GRC), find slaves automatically if such a thing can be unambiguous.  This is still not fully inferred slaves, it is inferred slaves when they are already present, but not indicated as slaves. (!373)(1566b9e9)
- **runtime**: enable `OCPI_APPLICATION_PARAMS` to work better and for more options like `dump` and `verbose` (still not documented). (!373)(1566b9e9)
- **runtime**: have our hdl sim drivers kill sims on `SIGTERM` as well as `SIGINT`. (!373)(1566b9e9)
- **tools**: `ocpigen`: make `ocpi_debug` a hidden property. (!373)(1566b9e9)
- **tools**: `ocpigr37`: make `ocpi_container` a pseudo-property for GRC to be "partly hidden". (!373)(1566b9e9)
- **tools**: add platform, worker, model, and "done" pseudo-properties to the `ocpigr37` output, and put them last after normal properties. (!373)(1566b9e9)
- **tools**: `ocpigen`: make `ocpi_endian` a hidden property. (!373)(1566b9e9)
- **tools**: add env var `OCPI_DISTRO_BUILD` to enable generic x86_64 distro build feature. (!380)(e849452b)

tools/scripts/ocpiadmin.sh
#!/bin/bash --noprofile
# This file is protected by Copyright. Please refer to the COPYRIGHT file
# distributed with this source distribution.
#
# This file is part of OpenCPI <http://www.opencpi.org>
#
# OpenCPI is free software: you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# OpenCPI is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#
# Originally, this script began life as "scripts/install-platform.sh".  Then,
# the functionality of "scripts/deploy-opencpi.sh" was integrated.  Now, both
# of the original scripts have gone away because it made no sense to have one
# integrated script with three names.  Unlike the originals (which were never
# exported, but it was discovered they needed to be), this script must be run
# with the OpenCPI environment properly initialized.

set -e

if [ "x$OCPI_CDK_DIR" = x ]
then
  #
  # OpenCPI environment not initialized: fatal error.
  #
  echo "$0: ERROR: OCPI_CDK_DIR not set!" >&2
  echo "From the top-level \"opencpi\" directory, try \"source cdk/opencpi-setup.sh -s\"." >&2
  exit 1
fi

#
# Must be in the OpenCPI base directory, at least until a
# clean way of dealing with relative paths can be implemented.
#
cd $OCPI_CDK_DIR/..

# Provides `setVarsFromMake`
source $OCPI_CDK_DIR/scripts/util.sh


function usage {
  if [ "$action" != install ]; then
    cat <<-EOF
To install a platform (by downloading it, if necessary, and building it):
  $(basename $0) install platform <platform> [-p PKG_ID [-u URL] [-g GIT_REV]]
To deploy a platform:
  $(basename $0) deploy platform <rcc_platform> <hdl_platform>
EOF
  else
    install_usage
  fi
  exit 1
}

function install_usage {
  cat <<-EOF
Usage: $(basename $0) install platform <platform> [-p PKG_ID [-u URL] [-g GIT_REV]]

Download, build, and register the built-in or remote OpenCPI RCC or HDL platform.

Required args:
  platform  Name of platform to install

Optional args:
  The -u and -g flags are only valid if the -p flag is specified as well. If
  the PKG_ID has already been downloaded and is detected, then PKG_ID can be
  omitted and just the PLATFORM name is required.

  -g GIT_REV, --git-revision GIT_REV
                        The branch, tag, or other valid git revision to checkout
                        after cloning the default git repo determined by PKG_ID.
                        GIT_REV defaults to the currently checked out OpenCPI
                        git revision.
  -p PKG_ID, --package-id PKG_ID
                        The OpenCPI Package ID that provides PLATFORM
  -u URL, --url URL     Use this URL when cloning the remote or local git repo
                        instead of the default url determined by PKG_ID
Examples:
  # xsim
  ocpiadmin install platform xsim

  # E31x
  ocpiadmin install platform e31x ocpi.osp.e3xx

  # PlutoSDR
  # PKG_ID not needed for second command as it has already been downloaded
  ocpiadmin install platform adi_plutosdr0_32 ocpi.osp.plutosdr
  ocpiadmin install platform plutosdr
EOF
  exit 1
}

function getvars {
  setVarsFromMake $OCPI_CDK_DIR/include/hdl/hdl-targets.mk ShellHdlTargetsVars=1
  setVarsFromMake $OCPI_CDK_DIR/include/rcc/rcc-targets.mk ShellRccTargetsVars=1
  if [ "$action" = deploy ]
  then
    export OCPI_ALL_RCC_PLATFORMS="$RccAllPlatforms" OCPI_ALL_HDL_PLATFORMS="$HdlAllPlatforms"
    return 0
  fi
  platforms="$RccAllPlatforms $HdlAllPlatforms"
  if isPresent $platform $platforms; then
    if isPresent $platform $RccAllPlatforms; then
	model=RCC
	v=RccPlatDir_$platform
    else
	model=HDL
	v=HdlPlatformDir_$platform
    fi
    platform_dir=`echo ${!v} | sed 's=^.*/projects/=./projects/='`
    return 0
  fi
  return 1
}

function bad {
  echo -e "Error: $@" >&2
  exit 1
}


#
# Quick and dirty argument parsing:
# getopt(s) overhead not required.
#
PARAMS=""
while (( "$#" )); do
  case "$1" in
    -g|--git-revision)
      if [ -n "$2" ] && [ ${2:0:1} != "-" ]; then
        GIT_REV=$2
        shift 2
      else
        bad "Argument for \"$1\" is missing"
      fi
      ;;
    -p|--package-id)
      if [ -n "$2" ] && [ ${2:0:1} != "-" ]; then
        PKG_ID=$2
        shift 2
      else
        bad "Argument for \"$1\" is missing"
      fi
      ;;
    -u|--url)
      if [ -n "$2" ] && [ ${2:0:1} != "-" ]; then
        URL=$2
        shift 2
      else
        bad "Argument for \"$1\" is missing"
      fi
      ;;

    # Undocumented flags
    -d|--distro)
      export OCPI_DISTRO_BUILD=1
      shift
      ;;
    -v|--verbose)
      verbose=-v
      shift
      ;;

    # Unsupported flags
    -*)
      HELP=1  # can't print usage yet as usage message is based on other args
      shift
      ;;

    # Preserve positional arguments
    *)
      PARAMS="$PARAMS \"$1\""
      shift
      ;;
  esac
done  # end parsing optional args and flags

# Set positional arguments in their proper place
eval set -- "$PARAMS"
unset PARAMS

# Parse and validate positional args.
action=$1

# Needs to be after $action is set as a different usage message is printed
# based on the action. Also, at least 3 positional arguments are required.
if [[ -n "$HELP" || $# -lt 3 ]]; then
  usage
fi

# $2 is always 'platform' for now
noun=$2
if [ "$noun" != platform ]; then
  bad "Unknown $action noun: '$noun'"
fi
if [ "$action" = install ]; then
  platform=$3
  if [ -z "$platform" ]; then
    bad 'Missing platform to install'
  fi
  # Check to ensure "PKG_ID" is non-null if either
  # "--url" or "--git-revision" were specified.
  if [[ ("$URL" || "$GIT_REV") && -z "$PKG_ID" ]]
  then
    bad 'PKG_ID is required if a URL or GIT_REV is specified'
  fi
elif [ "$action" = deploy ]; then
  rcc_platform=$3
  hdl_platform=$4
  if [ -z "$hdl_platform" ]; then
    bad 'Cannot deploy platform, missing required rcc and/or hdl platform'
  fi
else
  bad "Unknown action: '$action'"
fi

# End parsing and validation of positional args

#
# The "deploy" case is trivial: "deploy-platform.sh"
# does the heavy lifting.  Note undocumented "verbose"
# option: set to "-v" for debugging.
#
if [ "$action" = deploy ]
then
  getvars
  $OCPI_CDK_DIR/scripts/deploy-platform.sh $verbose $rcc_platform $hdl_platform
  exit $?
fi


if getvars; then
    echo The $model platform \"$platform\" is already defined in this installation, in $platform_dir.
    project_dir=$(echo $platform_dir | sed -e 's=/exports/=/=' -e "s=/.../platforms/$platform==" -e 's=/lib$==')
    if [ -n "$PKG_ID" ]; then
	echo The supplied project package-id for this platform, \"$PKG_ID\", will be ignored.
    fi
else
    echo The platform \"$platform\" is not defined in this installation yet.
    if [ -z "$PKG_ID" ]; then
	echo ERROR: no project package-id was specified, and
	echo platform \"$platform\" is not in a built-in project.
	echo Either the platform name is misspelled or you must supply a project package-id.
	exit 1
    fi
    project_dir=projects/osps/$PKG_ID
    if [ -d "$project_dir" ]; then
	echo There is already a directory at $project_dir, so it is assumed to contain the right project.
	echo It will be used and registered.  To force a new download of the project, unregister it and remove it.
        echo It will not be checked out for any particular tag or branch since it is assumed that it has
	echo either been done manually or previously by this script.
    else
	if [ -z "$URL" ]; then
	    URL=https://gitlab.com/opencpi/osp/$PKG_ID.git
	    echo No URL was specified, so it will default to the OpenCPI repo site: $URL
	fi
	echo "Downloading (git cloning) from $URL..."
	if git clone --no-checkout $URL $project_dir && test -d $project_dir; then
	    echo "Download/clone successful into $project_dir."
	else
	    echo "ERROR: download/clone of project \"$PKG_ID\" for platform \"$platform\" failed."
	    rm -r -f $project_dir
	    exit 1
	fi
	# Perform the checkout of the project repo
	# First assume we want the same tag/branch as the main repo.
	# If framework is on a tag, see if this repo is also.

	# Before registering it we will test whether it seems to be a working repo
	if tag=$(git describe --exact-match --tags 2>/dev/null); then
	    echo "The OpenCPI (framework) repo is on tag: $tag".
	    if branch=$(git branch --contains $(git rev-parse tags/$tag)); then
		echo "The branch for tag $tag is $branch";
	    else
		echo "ERROR: no git branch found for tag $tag"
		exit 1
	    fi
	elif branch=$(git branch --contains); then
	    branch=${branch#* }
	    echo "The OpenCPI (framework) repo is on branch: $branch."
	else
	    echo "ERROR: cannot determine branch of OpenCPI repo"
	    exit 1
	fi
	if [ -n "$GIT_REV" ]; then
	    echo Checking out the OSP at $project_dir using the user-supplied branch/tag: $GIT_REV
	    if (cd $project_dir && git checkout $GIT_REV); then
		echo The OSP at $project_dir checked out for branch/tag: $GIT_REV
	    else
		echo ERROR: failed to checkout the OSP for branch/tag \"$GIT_REV\".
		exit 1
	    fi
	elif [ -n "$tag" ]; then
	    if (cd $project_dir && git checkout $tag); then
		echo The OSP at $project_dir has been checked out to tag: $tag
	    else
		echo The OSP at $project_dir did not successfully checkout for tag: $tag.
		echo Checking out to branch $branch will be tried.
		if (cd $project_dir && git checkout $branch); then
		    echo The OSP at $project_dir checked out for branch: $branch
		else
		    echo ERROR: failed to checkout the OSP for either tag \"$tag\" or branch \"$branch\".
		    exit 1
		fi
	    fi
	else
	    echo Checking out the OSP at $project_dir for branch $branch.
	    if (cd $project_dir && git checkout $branch); then
		echo The OSP at $project_dir checked out for branch: $branch
	    else
		echo ERROR: failed to checkout the OSP for branch \"$branch\".
		exit 1
	    fi
	fi
    fi
    # Check if the platform is now available, without registering it
    export OCPI_PROJECT_PATH=`pwd`/$project_dir
    if getvars; then
	echo The $model platform \"$platform\" found after using the new project \"$PKG_ID\".
    else
	echo ERROR: in the downloaded project \"$PKG_ID\", at $project_dir, platform \"$platform\" is not visible.
	echo If you want to download it again, you must remove that directory and its contents.
	exit 1
    fi
    if [ ! -d "project-registry/$PKG_ID" ]; then
	if ocpidev register project $project_dir; then
	    echo The OSP at $project_dir has now been registered.
	else
	    echo ERROR: the project \"$PKG_ID\", at $project_dir, cannot be registered.
	    echo If you want to download it again, you must remove that directory and its contents.
	    exit 1
	fi
    fi
fi
if [ "$model" = RCC ]; then
    ./scripts/install-opencpi.sh $platform || exit 1
else
    ocpidev -d projects/core build --hdl --hdl-platform=$platform
    ocpidev -d projects/platform build --hdl --hdl-platform=$platform --no-assemblies
    ocpidev -d projects/assets build --hdl --hdl-platform=$platform --no-assemblies
    ocpidev -d projects/assets_ts build --hdl --hdl-platform=$platform --no-assemblies
    ocpidev -d projects/tutorial build --hdl --hdl-platform=$platform --no-assemblies

    # Make sure that tutorials can run after installation, note will do rcc too.
    [ "$platform" = xsim ] && ocpidev -d projects/tutorial build --hdl-platform=$platform

    # If project dir is not one of the core projects, build the platform
    if [[ -n "$platform_dir" && "$platform_dir" != *"/projects/core/"* \
          && "$platform_dir" != *"/projects/platform/"* \
          && "$platform_dir" != *"/projects/assets/"* ]]
    then
	ocpidev -d $project_dir build --hdl --hdl-platform=$platform --no-assemblies
	echo "HDL platform \"$platform\" built for OSP in $project_dir, including assemblies."
    fi
    ocpidev -d projects/assets build --hdl-platform=$platform hdl assembly testbias
    echo "HDL platform \"$platform\" built, with one HDL assembly (testbias) built for testing."
    echo "Preparing exported files for using this platform."
    #
    # At this point, we have an issue applicable to OSPs that have not
    # been previously installed.  A previous "getvars" call (above) sets
    # "platform_dir" to
    #   "./projects/osps/$PKG_ID/hdl/platforms/<hdl_platform>"
    # because
    #   "./projects/osps/$PKG_ID/hdl/platforms/<hdl_platform>/lib"
    # does not exist until the above "ocpidev" commands have been run,
    # i.e., this is a bootstrapping issue.  "platform_dir" can be updated
    # here by calling "getvars" one more time (after everything is built),
    # or in the "export-platform-to-framework.sh" script.
    #
    # No need to check the return value from "getvars" at this point.
    #
    getvars
    $OCPI_CDK_DIR/scripts/export-platform-to-framework.sh -v hdl $platform $platform_dir
fi
echo "Platform installation (download and build) for platform \"$platform\" succeeded."
exit 0
